src/provider/provider_tracking.c

```
@@ -7,73 +7,219 @@
                 size);
         return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
     }
 
     value->pool = pool;
     value->size = size;
+    value->n_children = 0;
 
-    int ret =
-        critnib_insert(hTracker->alloc_segments_map, (uintptr_t)ptr, value, 0);
-
+    assert(level < MAX_LEVELS_OF_ALLOC_SEGMENT_MAP);
+    int ret = critnib_insert(hTracker->alloc_segments_map[level],
+                             (uintptr_t)ptr, value, 0);
     if (ret == 0) {
-        LOG_DEBUG(
-            "memory region is added, tracker=%p, ptr=%p, pool=%p, size=%zu",
-            (void *)hTracker, ptr, (void *)pool, size);
+        LOG_DEBUG("memory region is added, tracker=%p, level=%i, pool=%p, "
+                  "ptr=%p, size=%zu",
+                  (void *)hTracker, level, (void *)pool, ptr, size);
         return UMF_RESULT_SUCCESS;
     }
+    if (ret == ENOMEM) {
+        umf_result = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
 
-    LOG_ERR("failed to insert tracker value, ret=%d, ptr=%p, pool=%p, size=%zu",
-            ret, ptr, (void *)pool, size);
+    LOG_ERR(
+        "failed to insert the tracker value: pool=%p, ptr=%p, size=%zu, ret=%d",
+        (void *)pool, ptr, size, ret);
 
     umf_ba_free(hTracker->alloc_info_allocator, value);
 
-    if (ret == ENOMEM) {
-        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    return umf_result;
+}
+
+static umf_result_t umfMemoryTrackerAdd(umf_memory_tracker_handle_t hTracker,
+                                        umf_memory_pool_handle_t pool,
+                                        const void *ptr, size_t size) {
+    assert(ptr);
+
+    umf_result_t umf_result = UMF_RESULT_ERROR_UNKNOWN;
+    tracker_alloc_info_t *parent_value = NULL;
+    tracker_alloc_info_t *rvalue = NULL;
+    uintptr_t parent_key = 0;
+    uintptr_t rkey = 0;
+    int parent_level = 0;
+    int level = 0;
+    int found = 0;
+
+    int ret = utils_mutex_lock(&hTracker->mutex);
```

It kills scalability

---

@vinser52 @bratpiorka @pbalcer @igchor @lplewa Thread Sanitizer and Valgrind (helgrind) fail without this lock and I do not know how make them pass without this lock:
https://github.com/ldorau/unified-memory-framework/actions/runs/13604042686/job/38033679491
https://github.com/ldorau/unified-memory-framework/actions/runs/13604042686/job/38033841818
https://github.com/ldorau/unified-memory-framework/actions/runs/13626003383/job/38083301826
https://github.com/ldorau/unified-memory-framework/actions/runs/13626003383/job/38083300612

---

@vinser52 this lock can be under `if (max_level > 1)` - what do you think?

---

I do not understand the scenario. Could you please explain it?
My understanding is that you are protecting against the case when the parent is concurrently removed while we are adding a child to the next layer, right? But I do not see how it is possible.

The only scenario I see is the following:

1. We have memory provider A and pool A on top of it.
2. Allocate buffer B from memory pool A and use it inside the Fixed provider.
3. Create memory pool B on top of the Fixed provider.
4. Thread 1 allocates memory from pool B. Internally pool B allocates from the Fixed provider and the tracking provider adds the corresponding memory block to the 1st layer as a child for buffer B (that is stored in tracker's 0-layer).
5. Thread 2 concurrently deallocates buffer B and removes it from the tracker.

But it is an invalid usage of the Fixed providers from the client's perspective. Do we have any valid scenario when race happens?

---

I will think more about it.
But anyway, what if there is no valid scenario? Thread Sanitizer and Valgrind (helgrind) fail without this lock anyway. How do you want to make those CI builds pass?

---

> Thread Sanitizer and Valgrind (helgrind) fail without this lock anyway. How do you want to make those CI builds pass?

Thread Sanitizer and Valgrind report potential issues. We (as software developers) should analyze them and decide whether it is a real issue or a false-positive. That is why I asked what is the real scenario. If there is a real issue than we should think about how to fix it. If there is no real issue and we consider it as a false positive we should suppress it in Valgrind and Sanitizers.

---

here is the failing scenario:
1) thread #1 tries to add entry to the tracker and loops through critnibs
2) thread #2 already removed an entry 
3) thread #3 allocates new entry (reuses memory freed by #2)

then there is a race: thread #1 reads memory just initialized by #3

---

@bratpiorka is it multilayer case or singlelayer?
If multilayer, I have described something similar in my previous [comment](https://github.com/oneapi-src/unified-memory-framework/pull/1143#discussion_r1977631629), but as I pointed it is invalid usage of UMF.

if singlelayer, than I did not get the use case.

---

> Thread Sanitizer and Valgrind report potential issues. We (as software developers) should analyze them and decide whether it is a real issue or a false-positive. That is why I asked what is the real scenario. If there is a real issue than we should think about how to fix it. If there is no real issue and we consider it as a false positive we should suppress it in Valgrind and Sanitizers.

That's obvious. I meant that - AFAIK - there are no suppressions in Sanitizers, only a few annotations, so it is easy to suppress such errors in Valgrind, but it can be very difficult or even impossible to make the Sanitizers builds pass.

---

First, we should answer if it is a real issue or a false-positive.
@ldorau I'm confused, are you saying it is false-positive? If so, I am strongly against introducing a lock just to suppress false-positive

---

> First, we should answer if it is a real issue or a false-positive. @ldorau I'm confused, are you saying it is false-positive? If so, I am strongly against introducing a lock just to suppress false-positive

No, I'm not saying it is false-positive. I am still investigating it.

---

Good, I am also looking into the code and investigating this issue.

---

I removed the lock and add atomics. Please re-review.

---

Lock removed. @vinser52 please resolve this issue.

---

See: https://github.com/oneapi-src/unified-memory-framework/pull/1143#issuecomment-2703620746

---

Done